chore: add doc for the field attribute (#3209)
noir-lang/noir: docs/docs/language_concepts/01_functions.md

```

See [Lambdas](./08_lambdas.md) for more details.

## Attributes

> Attribute is metadata that can be applied to a function, using the following syntax: `#[attribute(value)]`.

Reviewer: ```suggestion
Attributes are metadata that can be applied to a function using the following syntax: `#[attribute(value)]`.
```

Reviewer: Before we merge, can we fix this grammar issue?

Reviewer: Nevermind, my diff was out of date. Looks good

Supported attributes include:
- **builtin**: the function is implemented by the compiler, for efficiency purposes.
- **deprecated**: mark the function as *deprecated*. Calling the function will generate a warning: `warning: use of deprecated function`

> - **field**: conditional compilation of code. See below for more details

Reviewer: ```suggestion
- **field**: Used to enable conditional compilation of code depending on the field size. See below for more details
```

Author: Done

- **oracle**: mark the function as *oracle*; meaning it is an external unconstrained function, implemented in noir_js. See [Unconstrained](./05_unconstrained.md) and [Noir js](../noir_js/noir_js.md) for more details.

> - **test**: mark the function as unit tests. See [Tests](../nargo/02_testing.md) for more details

Reviewer: Can we re-order this attribute section? I think it'd be better if attributes were ordered by their importance to end users. E.g. 1. test, 2. field, 3. deprecated, 4. builtin, and 5. oracle

### Field Attribute
The field attribute defines which field the function is compatible for. The function is conditionally compiled, under the condition that the field attribute matches the Noir native field.

> The field can be defined implicitely, by using the name of the elliptic curve usually associated to it - for instance bn254, bls12_381 - or explicitely by using the field (prime) order, in decimal or hexadecimal form.

Reviewer: ```suggestion
The field can be defined implicitly, by using the name of the elliptic curve usually associated to it - for instance bn254, bls12_381 - or explicitly by using the field (prime) order, in decimal or hexadecimal form.
```

Author: Done

> The field can be defined implicitely, by using the name of the elliptic curve usually associated to it - for instance bn254, bls12_381 - or explicitely by using the field (prime) order, in decimal or hexadecimal form.

Reviewer 2: ```suggestion
The field can be defined implicitly, by using the name of the elliptic curve usually associated to it - for instance bn254, bls12_381 - or explicitly by using the field (prime) order, in decimal or hexadecimal form.
```

Author: Done

> As a result, it is possible to define multiple times the function with a different field attribute. This is usefull for instance when a function requires different parameters depending on the underlying elliptic curve.

Reviewer: ```suggestion
As a result, it is possible to define multiple versions of a function with each version specialized for a different field attribute. This can be useful when a function requires different parameters depending on the underlying elliptic curve.
```

Author: Done

> As a result, it is possible to define multiple times the function with a different field attribute. This is usefull for instance when a function requires different parameters depending on the underlying elliptic curve.

Reviewer 2: ```suggestion
As a result, it is possible to define a function multiple times with different field attributes. This is useful, for instance, when a function requires different parameters depending on the underlying elliptic curve.
```

Author: I used Jake's version, but I can use yours if you prefer.

> Example: we define the function `foo()` three times below. One for the default noir bn254 curve, one for the field $\mathbb F_{23}$, which will normally never be used by Noir, and the bls12_381 curve.

Reviewer: ```suggestion
Example: we define the function `foo()` three times below. Once for the default Noir bn254 curve, once for the field $\mathbb F_{23}$, which will normally never be used by Noir, and once again for the bls12_381 curve.
```

```rust
#[field(bn254)]
fn foo() -> u32 {
    1
}

#[field(23)]
fn foo() -> u32 {
    2
}

// This commented code would not compile as foo would be defined twice because it is the same field as bn254
// #[field(21888242871839275222246405745257275088548364400416034343698204186575808495617)]
// fn foo() -> u32 {
//     2
// }

#[field(bls12_381)]

> fn foo() -> u32 {
>     3

Reviewer: ```suggestion
fn foo() -> u32 {
    3
}
```

Author: Done

```

> If the name is not known to Noir, it will discard the function. Field names are case insensitive.

Reviewer: ```suggestion
If the field name is not known to Noir, it will discard the function. Field names are case insensitive.
```

Author: Done
